File: src/year2021/month6/no207/CourseSchedule.java
package year2021.month6.no207;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class CourseSchedule {

    public static void main(String[] args) {
        int numCourses1 = 2;
        int[][] prerequisites1 = {{1, 0}};
        int numCourses2 = 2;
        int[][] prerequisites2 = {{1, 0}, {0, 1}};
        System.out.println(canFinish2(numCourses1, prerequisites1));
        System.out.println(canFinish2(numCourses2, prerequisites2));
    }

    private static boolean canFinish2(int numCourses, int[][] prerequisites) {
        // DFS, 如果一个有向图可以生成拓扑排序等价于有向图无环
        // 则使用 DFS 对访问的路径进行标记，若 DFS 可以将所有节点都访问完
        // 则代表无环，否则碰到先前访问的标记，则代表有环
        // time is O(N + E), space is O(N + E)
        List<List<Integer>> adjacentTable = new ArrayList<>(numCourses);
        for (int i = 0; i < numCourses; i++) {
            adjacentTable.add(new LinkedList<>());
        }
        for (int[] prerequisite : prerequisites) {  //初始化邻接表
            int from = prerequisite[1];
            int to = prerequisite[0];
            adjacentTable.get(from).add(to);
        }
        int[] visited = new int[numCourses];    //标记是否访问，-1 未访问，0 正在访问，1 已访问
        Arrays.fill(visited, -1);           //初始化为 -1 未访问
        for (int i = 0; i < numCourses; i++) {
            if (visited[i] != -1) {
                continue;
            }
            if (dfs(adjacentTable, visited, i)) {   // dfs判断是否有环
                return false;
            }
        }
        return true;
    }

    private static boolean dfs(List<List<Integer>> adjacentTable, int[] visited, int cur) {
        visited[cur] = 0;   // 标记为 0 正在访问
        List<Integer> nextNodes = adjacentTable.get(cur);
        for (int next : nextNodes) {
            if (visited[next] == 1) {
                continue;
            }
            if (visited[next] == 0) {   // 有环
                return true;
            }
            if (dfs(adjacentTable, visited, next)) {
                return true;
            }
        }
        visited[cur] = 1;   // 标记为访问完成
        return false;
    }

    private static boolean canFinish1(int numCourses, int[][] prerequisites) {
        // 入度表 + 邻接表 + 拓扑排序 + BFS, time is O(N + E), space is O(N + E)
        int[] inDegrees = new int[numCourses];
        List<Integer>[] adjacence = new LinkedList[numCourses];
        for (int[] prerequisite : prerequisites) {
            int from = prerequisite[1];
            int to = prerequisite[0];
            inDegrees[to]++;
            if (adjacence[from] == null) {
                adjacence[from] = new LinkedList<>();
            }
            adjacence[from].add(to);
        }
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < inDegrees.length; i++) {
            if (inDegrees[i] == 0) {
                queue.offer(i);
            }
        }
        int count = 0;
        // BFS
        while (!queue.isEmpty()) {
            Integer node = queue.poll();
            count++;
            List<Integer> next = adjacence[node];
            if (next != null) {
                next.forEach(item -> {
                    inDegrees[item]--;
                    if (inDegrees[item] == 0) {
                        queue.offer(item);
                    }
                });
            }
        }
        return count == numCourses;
    }

    public static boolean canFinish(int numCourses, int[][] prerequisites) {
        // 入度表 + 邻接表 + 拓扑排序 + 暴力扫描入度表
        int[] inDegrees = new int[numCourses];
        List<Integer>[] adjacence = new LinkedList[numCourses];
        for (int[] prerequisite : prerequisites) {
            int from = prerequisite[1];
            int to = prerequisite[0];
            inDegrees[to]++;
            if (adjacence[from] == null) {
                adjacence[from] = new LinkedList<>();
            }
            adjacence[from].add(to);
        }
        boolean flag = true;
        while (flag) {
            flag = false;
            for (int i = 0; i < inDegrees.length; i++) {
                if (inDegrees[i] != 0) {
                    continue;
                }
                flag = true;
                List<Integer> next = adjacence[i];
                if (next != null) {
                    next.forEach(index -> inDegrees[index]--);
                }
                inDegrees[i] = -1; //标记为已访问
            }
        }
        return Arrays.stream(inDegrees).boxed().allMatch(num -> num < 0);
    }

}

/*
* 你这个学期必须选修 numCourses 门课程，记为 0 到 numCourses - 1 。

在选修某些课程之前需要一些先修课程。 先修课程按数组 prerequisites 给出，其中 prerequisites[i] = [ai, bi] ，表示如果要学习课程 ai 则 必须 先学习课程  bi 。

例如，先修课程对 [0, 1] 表示：想要学习课程 0 ，你需要先完成课程 1 。
请你判断是否可能完成所有课程的学习？如果可以，返回 true ；否则，返回 false 。

 

示例 1：

输入：numCourses = 2, prerequisites = [[1,0]]
输出：true
解释：总共有 2 门课程。学习课程 1 之前，你需要完成课程 0 。这是可能的。
示例 2：

输入：numCourses = 2, prerequisites = [[1,0],[0,1]]
输出：false
解释：总共有 2 门课程。学习课程 1 之前，你需要先完成 课程 0 ；并且学习课程 0 之前，你还应先完成课程 1 。这是不可能的。
 

提示：

1 <= numCourses <= 105
0 <= prerequisites.length <= 5000
prerequisites[i].length == 2
0 <= ai, bi < numCourses
prerequisites[i] 中的所有课程对 互不相同

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/course-schedule
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
* */

File: go/year2025/month3/no207/CourseSchedule.go
package main

func main() {
	numCourses1 := 2
	prerequisites1 := [][]int{{1, 0}}
	println(canFinish(numCourses1, prerequisites1)) //true
	numCourses2 := 2
	prerequisites2 := [][]int{{1, 0}, {0, 1}}
	println(canFinish(numCourses2, prerequisites2)) //false
}

// BFS，times: O(n+e), space: O(n+e)
// 思路：邻接表遍历，每次找到入度为 0 的节点，将其加入队列，然后将其指向的节点的入度减 1，直到队列为空，如果遍历完所有节点，则无环
func canFinish(numCourses int, prerequisites [][]int) bool {
	// 构建邻接表 from -> to list
	adjacentTables := make([][]int, numCourses)
	for i := 0; i < numCourses; i++ { // O(n)
		adjacentTables[i] = make([]int, 0)
	}
	// 构建入度表	to -> inDegree
	inDegrees := make([]int, numCourses)
	for _, prerequisite := range prerequisites { // O(e)
		from, to := prerequisite[1], prerequisite[0]
		adjacentTables[from] = append(adjacentTables[from], to)
		inDegrees[to]++
	}
	// 入度为 0 的节点加入队列
	queue := make([]int, 0)
	for i, degree := range inDegrees { // O(n)
		if degree == 0 {
			queue = append(queue, i)
		}
	}
	// visited 记录访问过的节点数
	visited := 0
	// BFS, 每个点每条边都只会访问一次，总复杂度 O(n+e)
	for len(queue) > 0 {
		nextQueue := make([]int, 0)
		for _, from := range queue {
			visited++ // 访问 from，节点数加 1，需要保证 queue 中的节点都是未访问的
			for _, to := range adjacentTables[from] {
				inDegrees[to]--         // 当前 from 已访问，因此 to 的入度都可以减 1
				if inDegrees[to] == 0 { // 入度为 0 的节点加入队列，这个条件可以保证 to 是未访问过的
					nextQueue = append(nextQueue, to)
				}
			}
		}
		queue = nextQueue // 下一轮访问的节点
	}
	return visited == numCourses
}

// DFS，times: O(n+e), space: O(n^2), n 为节点数，e 为边数
// 思路：如果一个有向图能生成拓扑排序，则等价于有向图无环，对每个节点进行 DFS，若所有节点都不在环内，则可以生成拓扑排序
func canFinish1(numCourses int, prerequisites [][]int) bool {
	// 构建邻接矩阵
	adjacentMatrix := make([][]bool, numCourses) // o(n)
	for i := 0; i < numCourses; i++ {
		adjacentMatrix[i] = make([]bool, numCourses)
	}
	for _, prerequisite := range prerequisites { // o(e)
		from, to := prerequisite[1], prerequisite[0]
		adjacentMatrix[from][to] = true
	}
	// 记录节点状态：0 未访问，1 正在访问，2 已访问
	visited := make([]int, numCourses)
	// 总复杂度 o(n+e)
	for i := 0; i < numCourses; i++ { // o(n)，每个节点最多访问一次
		if hasLoop(adjacentMatrix, visited, i) { // o(e)，每条边最多访问一次
			return false
		}
	}
	return true
}

// DFS，判断是否有环
func hasLoop(adjacentMatrix [][]bool, visited []int, i int) bool {
	if visited[i] == 2 { // 已访问，就不在访问了，避免重复访问造成死循环
		return false
	}
	if visited[i] == 1 { // 正在访问，说明有环
		return true
	}
	visited[i] = 1
	for next, canAccess := range adjacentMatrix[i] {
		if canAccess && hasLoop(adjacentMatrix, visited, next) {
			return true
		}
	}
	visited[i] = 2
	return false
}

/*
 你这个学期必须选修 numCourses 门课程，记为 0 到 numCourses - 1 。
在选修某些课程之前需要一些先修课程。 先修课程按数组 prerequisites 给出，其中 prerequisites[i] = [ai, bi] ，表示如果要学习课程 ai 则 必须 先学习课程  bi 。
例如，先修课程对 [0, 1] 表示：想要学习课程 0 ，你需要先完成课程 1 。
请你判断是否可能完成所有课程的学习？如果可以，返回 true ；否则，返回 false 。

示例 1：
输入：numCourses = 2, prerequisites = [[1,0]]
输出：true
解释：总共有 2 门课程。学习课程 1 之前，你需要完成课程 0 。这是可能的。

示例 2：
输入：numCourses = 2, prerequisites = [[1,0],[0,1]]
输出：false
解释：总共有 2 门课程。学习课程 1 之前，你需要先完成​课程 0 ；并且学习课程 0 之前，你还应先完成课程 1 。这是不可能的。


提示：
1 <= numCourses <= 2000
0 <= prerequisites.length <= 5000
prerequisites[i].length == 2
0 <= ai, bi < numCourses
prerequisites[i] 中的所有课程对 互不相同
*/
